doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,697 @@
+Features including distributed arrays, implicit communication, and global
+view make it easier to go from a serial program, to a parallel program,
+to a distributed-memory parallel program. However, these same features
+can also create performance problems. It takes additional work to go from
+a distributed-memory parallel program to one that performs well. This
+document provides a guide of sorts along the way. It will highlight the
+issues that one needs to be aware of when optimizing the performance of a
+Chapel program and talk about how the performance issues can be resolved.
+
+.. note::
+
+   This document is focused on how Chapel users can improve the
+   performance of their applications. There are many potential areas for
+   improvement in the compiler, runtime, and standard library that can
+   also improve the performance of Chapel programs. In fact, some
+   optimizations to the Chapel programming system have made benchmarks
+   run many times faster without needing any change to the source code at
+   all! These are important techniques as well, but they are out of scope
+   for this document.
+
+.. note::
+
+   This document is a draft and a work in progress. Updates to it based
+   on experience with optimization are welcome!
+
+
+A Few Words About Process
+-------------------------
+
+Before we dive into optimization techniques, let's briefly discuss some
+good programming practices when working with Chapel. Optimization efforts
+will involve changing the code, and optimizations that cause the program
+to stop working correctly aren't any use at all. It needs to be easy to
+check that everything is working.
+
+Revision Control
+~~~~~~~~~~~~~~~~
+
+It's important to use revision control tools (e.g. ``git``). Revision
+control helps with optimization specifically:
+
+ * You can go back to an earlier version if your changes break the
+   program or make it slower
+
+ * You can record your progress towards improving performance
+
+Testing
+~~~~~~~
+
+It's important to have tests that are easy to run in order to check that
+your program is working correctly.
+
+We have used the following steps when testing a new change to a parallel
+and distributed Chapel application:
+
+ * Compile with checking enabled (i.e. without ``--fast``)
+
+   * Test with 1 locale and ``--dataParTasksPerLocale=1``
+
+   * Gradually increase ``--dataParTasksPerLocale`` until you are
+     satisfied it is working. At a minimum, try leaving this argument off
+     to run with number of tasks == number of cores.
+
+   * Test while simulating multiple locales on your development system
+     (see the next section)
+
+     Gradually increase the number of locales and the number of tasks
+     until you are satisfied it is working.
+
+You can compile with ``--fast`` and measure performance before or after
+this process. If you are targeting distributed memory, measure
+performance locally oversubscribed (see below) or by running on a
```

Comment:
Are you recommending perf-testing oversubscribed gasnet here? I recommend against it, rather strongly.

Reply:
I've changed it to this

> You can compile with ``--fast`` and measure performance before or after this process. If you are targeting distributed memory, it's ideal to test performance on a multi-node system. Occasionally, it's useful to simulate multiple locales on a laptop (described below). Be aware that the performance of such a simulation won't necessarily match the real multi-node system.